### PR TITLE
#791@patch: Fix 🔧 #788 #789 #791 and #795.

### LINKED ISSUE #788: xmlhttprequest is missing response getter
```javascript
xhr.open("GET", "http://google.com", true);
xhr.responseType = "blob";
xhr.send();

xhr.onload = function () {
	console.log("onload", this.response);
};
// "onload" undefined
```

### LINKED ISSUE #789: xmlhttprequest is missing loading state in sync request.
https://github.com/capricorn86/happy-dom/blob/7f8ee017d6e1c395eb3976fae094cda0f7a6c499/packages/happy-dom/src/xml-http-request/XMLHttpRequest.ts#L570-L574

### LINKED ISSUE #791: xmlhttprequest header definition error
header is object.

https://github.com/capricorn86/happy-dom/blob/7f8ee017d6e1c395eb3976fae094cda0f7a6c499/packages/happy-dom/src/xml-http-request/XMLHttpRequest.ts#L73

https://github.com/capricorn86/happy-dom/blob/7f8ee017d6e1c395eb3976fae094cda0f7a6c499/packages/happy-dom/src/xml-http-request/XMLHttpRequest.ts#L980-L984

### LINKED ISSUE #795: xmlhttprequest error progress bugs
Error handling causes the program to run incorrectly.
needs insert `return` statement to stop.

https://github.com/capricorn86/happy-dom/blob/7f8ee017d6e1c395eb3976fae094cda0f7a6c499/packages/happy-dom/src/xml-http-request/XMLHttpRequest.ts#L834

### PR DESCRIPTION
**Pull request by Mas0nShi**
Resolves #788, resolves #789, resolves #791, resolves #795.

Others:
- complete local request response (async/sync).

### CODE REVIEW

File: packages/happy-dom/src/xml-http-request/utilities/XMLHttpRequestURLUtility.ts [hunk 1]
```diff
@@ -1,4 +1,78 @@
...
+	bin: 'application/octet-stream',
+	doc: 'application/msword',
+	docx: 'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
+	xls: 'application/vnd.ms-excel',
+	xlsx: 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
+	ppt: 'application/vnd.ms-powerpoint',
+	pptx: 'application/vnd.openxmlformats-officedocument.presentationml.presentation',
+	odt: 'application/vnd.oasis.opendocument.text',
+	ods: 'application/vnd.oasis.opendocument.spreadsheet'
+};
```
Comment (btea) [inline packages/happy-dom/src/xml-http-request/utilities/XMLHttpRequestURLUtility.ts:75]: `mime-type` may use this library https://github.com/broofa/mime.

Comment (Mas0nShi) [reply packages/happy-dom/src/xml-http-request/utilities/XMLHttpRequestURLUtility.ts:75]: well, I'm agree with you. 
Inevitably, this adds another third-party package, what do you think? @capricorn86

**Review by anubhavagwl — DISMISSED** [review]

File: packages/happy-dom/src/xml-http-request/utilities/XMLHttpRequestURLUtility.ts (hunk 1 above)
Comment (capricorn86) [reply packages/happy-dom/src/xml-http-request/utilities/XMLHttpRequestURLUtility.ts:75]: I have a little bit of a bad experience with third-party packages. My suggestion is to leave it as you did it for now.

There are also plans to support both ESM soon (with a fallback for CommonJS), and third party packages can sometimes cause problems with ESM.

**Review by capricorn86 — APPROVED** [review]
Thank you for your contribution! :slightly_smiling_face:

**Comment by capricorn86** [comment]
Is there any way we can verify this with some unit tests?

File: packages/happy-dom/src/xml-http-request/utilities/XMLHttpRequestURLUtility.ts (hunk 1 above)
Comment (capricorn86) [reply packages/happy-dom/src/xml-http-request/utilities/XMLHttpRequestURLUtility.ts:75]: I have a little bit of a bad experience with third party packages. My suggestion is to leave it as you did it.

**Comment by Mas0nShi** [comment]
> Is there any way we can verify this with some unit tests?

I would argue that three PR's other than #788 (#789 #791 #795) would struggle to have a clear test unit.

File: packages/happy-dom/src/xml-http-request/XMLHttpRequest.ts [hunk 2]
```diff
@@ -146,6 +146,15 @@ export default class XMLHttpRequest extends XMLHttpRequestEventTarget {
...
 
+	/**
+	 * Returns the response.
+	 *
+	 * @returns Response.
+	 */
+	public get response(): ArrayBuffer | Blob | IDocument | object | string {
+		return this._state.response;
+	}
+
```
Comment (Mas0nShi) [inline packages/happy-dom/src/xml-http-request/XMLHttpRequest.ts:157]: as you see, it's easy to add test unit for #788 

```suggestion
	/**
	 * Returns the response.
	 *
	 * @returns Response.
	 */
	public get response(): ArrayBuffer | Blob | IDocument | object | string {
		return this._state.response;
	}

```

File: packages/happy-dom/src/xml-http-request/XMLHttpRequest.ts [hunk 3]
```diff
@@ -597,7 +607,11 @@ export default class XMLHttpRequest extends XMLHttpRequestEventTarget {
 			};
 			this._state.status = response.statusCode;
 			this._state.statusText = response.statusMessage;
+			// Although it will immediately be set to loading,
+			// According to the spec, the state should be headersRecieved first.
+			this._setState(XMLHttpRequestReadyStateEnum.headersRecieved);
 			// Sync responseType === ''
+			this._setState(XMLHttpRequestReadyStateEnum.loading);
```
Comment (Mas0nShi) [inline packages/happy-dom/src/xml-http-request/XMLHttpRequest.ts:614]: but in this #789, it is difficult to implement the test unit due to implementation problems that make this part of the change happen "in a flash" and almost negligible.

File: packages/happy-dom/src/xml-http-request/XMLHttpRequest.ts [hunk 4]
```diff
@@ -831,10 +845,14 @@ export default class XMLHttpRequest extends XMLHttpRequestEventTarget {
 			data = await FS.promises.readFile(decodeURI(url.pathname.slice(1)));
 		} catch (error) {
 			this._onError(error);
+			// Release async task.
+			this._ownerDocument.defaultView.happyDOM.asyncTaskManager.endTask(this._state.asyncTaskID);
+			return;
```
Comment (Mas0nShi) [inline packages/happy-dom/src/xml-http-request/XMLHttpRequest.ts:850]: #795

**Comment by capricorn86** [comment]
@Mas0nShi again, thank you for contribution! :star2: 

Sorry for taking such a long time before looking into this PR again.

I have added unit tests and also fixed a problem with the logic for detecting MIME type which one of the unit tests caught.

**Pull request merged**